pytorch_lightning/plugins/ddp_sequential_plugin.py

```diff
@@ -0,0 +1,403 @@
+        This is turned on by default and can be turned off via the checkpoint argument.
+
+        You should determine the balance when defining the plugin,
+        or you can pass an example input array via the LightningModule to infer a balance.
+        The module will be partitioned into multiple devices according to the given balance. You may also rely on
+        your own heuristics to find your own optimal configuration.
+
+        Args:
+            balance: The balance of the model, i.e `[2, 2]` (two layers on each GPU).
+                If not provided assumes user provides an input example array to find a balance on all GPUs.
+            microbatches: Allows for parallelization to reduce device utilization
+                by splitting the batch into further smaller batches.
+            checkpoint: Enables gradient checkpointing. Options: `['always', 'except_last', 'never']`
+            balance_mode: Type of balance heuristic to use if balance to be inferred.
+            'balance_by_size': checks memory usage of each layer and determines balance,
+            'balance_by_time': checks time of each layer and determines balance
+            pipelined_backward: if True, call ``torch.autograd.backward`` once per microbatch on the
+                backward pass (instead of once for the whole batch). This works
+                around a potential deadlock in pytorch when using tensor parallelism
+                at the same time. Defaults to `True` if `get_model_parallel_world_size() > 1`
+                (default: `None`)
+        """
+        self._check_pipe_available()
+        super().__init__(**kwargs)
+
+        self.balance = balance
+
+        self.microbatches = microbatches
+        self.checkpoint = checkpoint
+        self.balance_mode = balance_mode
+        self.pipelined_backward = pipelined_backward
+        self.main_rpc_process = False  # Updated by main process, default for all secondary processes
+
+    def init_ddp_connection(
+            self,
+            trainer,
+            cluster_environment,
+            global_rank: int,
+            world_size: int,
+            is_slurm_managing_tasks: bool = True,
+    ) -> None:
+        trainer.prepared_for_backwards = False
+        self._check_arguments(trainer)
+        if self._skip_init_connections(trainer):
+            return
+        super().init_ddp_connection(
+            trainer=trainer,
+            cluster_environment=cluster_environment,
+            global_rank=global_rank,
+            world_size=world_size,
+            is_slurm_managing_tasks=is_slurm_managing_tasks
+        )
+        super().init_rpc_connection(
+            global_rank=global_rank,
+            world_size=world_size
+        )
+        self.gpus_per_model = self._infer_check_num_gpus(trainer)
+        self.init_model_parallel_groups(trainer)
+        self.set_main_rpc_process()
+
+        self._check_sequential_model_exists(trainer)
```

Comment:
why we cannot pass model as trainer gest model, hopefully, `self` anyway...
probably a similar case for other guys here too, you name method action with the model but you pass trainer...
or let's allow passing trainer or model
```
def _check_sequential_model_exists(self, model=None, trainer=None):
    assert model or trainer
    ...
````

Reply:
done